[Sema] Delay 'explicit self required' check until closure has valid type (SR-13273)

Jumhyn opened the pull request:
Avoid diagnosing "explicit self required" errors before we have the
type for a closure, since such closures will never be known-noescape.

Resolves SR-13273.

theblixguy commented on test/expr/closure/closures.swift:
```diff
@@ -504,3 +504,19 @@ var qux1: (() -> Void)? = {}
 SR9839(forceUnwrap(qux1))
+
+// SR-13273
+class SR13273 {
+    var y: Int = 0
+
+    func foo() {
+        bar {
+            bar { [y] in // OK
+                print(y)
+            }
+        }
+    }
+
+    func bar(_: () -> Void) {}
+}
```
Nit: Could you adjust the indentation here to match the rest of the file?

theblixguy commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1487,10 +1487,15 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      // If we don't have a type for this closure yet, don't offer diagnostics
+      // for required 'self'. We will have another chance when (if) the closure
+      // successfully typechecks.
+      if (!AFR.hasType())
```
You might also want to check whether the type has an error.

theblixguy commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1487,10 +1487,15 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
                   const AbstractClosureExpr *CE) {
+      auto AFR = AnyFunctionRef(const_cast<AbstractClosureExpr *>(CE));
+      // If we don't have a type for this closure yet, don't offer diagnostics
+      // for required 'self'. We will have another chance when (if) the closure
+      // successfully typechecks.
+      if (!AFR.hasType())
+        return false;
       // If the closure's type was inferred to be noescape, then it doesn't
       // need qualification.
-      return !AnyFunctionRef(const_cast<AbstractClosureExpr *>(CE))
-               .isKnownNoEscape();
+      return !AFR.isKnownNoEscape();
```
I don't think it's really necessary to `const_cast` and wrap in `AnyFunctionRef` to access this info, if you're already doing the `hasType()` or `getType()->hasError()` checks here. I think you can just do:

```c++
auto closureTy = CE->getType();
if (!closureTy || closureTy->hasError())
  return false;
return !closureTy->castTo<AnyFunctionType>()->isNoEscape();
```

Jumhyn replied:
No I think that makes sense. I think you're right that `isKnownNoEscape` is just the wrong thing to check here, since what we really want is "`isKnownPotentiallyEscaping`". Since `isKnownNoEscape` is used elsewhere I think your suggested adjustment is a good way to go.

Jumhyn commented:
I suspect that the just-merged #33116 addresses SR-13273 as well, but need to verify.

Jumhyn commented:
Confirmed that #33116 addresses the bug this PR was meant to fix. However, the logic here still seems like an improvement over what exists currently. Perhaps it is independently worthwhile to correct the logic here, and provide an additional regression test for the capture list behavior?

theblixguy commented:
Yeah, that sounds good to me.

varungandhi-apple commented:
Couple of workflow related suggestions:

1. Please use more descriptive commit messages and PR titles. Perhaps it could be `Delay diagnosis of "explicit self required" error` or `Avoid diagnosing "explicit self required" error eagerly`  something similar. The SR number can go into parens (in the commit/PR title), or otherwise mentioned in the commit message, but it shouldn't be the only thing in the title.
2. Prefer rebasing over merging, to avoid unnecessary commits polluting the history.

Jumhyn commented:
@varungandhi-apple Noted, thank you! The commit message had a more descriptive body, but I've elaborated on the commit/PR title and rebased onto master.

theblixguy commented:
@swift-ci please smoke test

DougGregor commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1491,10 +1491,15 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      // If we don't have a type for this closure yet (or the closure has error
+      // type), don't offer diagnostics for required 'self'. We will have
+      // another chance when (if) the closure successfully typechecks.
+      if (!closureTy || closureTy->hasError())
```
Do we need the `!closureTy`? I thought this code wouldn't be run at all if something failed.

Jumhyn replied:
Before you merged #33116, the 'explicit `self`' check here was getting triggered before we had resolved the types of all containing closures. So AFAICT there's nothing actually preventing this check from happening before we have a type, other than ensuring `performSyntacticExprDiagnostics` doesn't get called before all containing closures have a type.

If you're satisfied that this is otherwise guaranteed somewhere up the stack, then yeah, I believe this check is redundant. Could also be changed to an assertion.

Jumhyn replied:
@DougGregor Sorry for the ping, just want to close the loop on this PR. If you have the time to provide your preference among the following options I'd appreciate it:

- Keep the runtime check as it is in this PR since we don't have 'up the stack' checking that the necessary types are resolved before calling into these diagnostics.
- Change this check to an assertion.
- Keep the logic in `isClosureRequiringSelfQualification` as it is on `master` and add the additional regression test.
- Close out this PR without merging since #33116 sufficiently addresses the underlying issue.

I'm happy with all of the above, just don't want to let this PR sit around indefinitely and clog up the list of PRs!

Jumhyn committed:
[Sema] Delay 'explicit self required' check until closure has valid type

Avoid diagnosing "explicit self required" errors before we have the
type for a closure, since such closures will never be known-noescape.

Fixes SR-13273.

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
